# third_party/RoarGraph
# upstream: https://github.com/matchyc/RoarGraph
# base:     78bf05cf248195007604d9e2386c82566a1818c2
# pinned:   c2145e04af59126d0ab7ba15eb4b907e60147d23
#
# Regenerate with: python scripts/check_patches.py --write
# Apply with:      git -C third_party/RoarGraph apply patches/roargraph-pyroar-bindings.patch

diff --git a/include/efanna2e/distance.h b/include/efanna2e/distance.h
index 263c33b..1440de6 100644
--- a/include/efanna2e/distance.h
+++ b/include/efanna2e/distance.h
@@ -37,55 +37,10 @@ class DistanceL2 : public Distance {
 
    public:
     float compare(const float *x, const float *y, unsigned d) const {
-        __m512 msum0 = _mm512_setzero_ps();
 
-        while (d >= 16) {
-            __m512 mx = _mm512_loadu_ps(x);
-            x += 16;
-            __m512 my = _mm512_loadu_ps(y);
-            y += 16;
-            const __m512 a_m_b1 = mx - my;
-            msum0 += a_m_b1 * a_m_b1;
-            d -= 16;
-        }
-
-        __m256 msum1 = _mm512_extractf32x8_ps(msum0, 1);
-        msum1 += _mm512_extractf32x8_ps(msum0, 0);
-
-        if (d >= 8) {
-            __m256 mx = _mm256_loadu_ps(x);
-            x += 8;
-            __m256 my = _mm256_loadu_ps(y);
-            y += 8;
-            const __m256 a_m_b1 = mx - my;
-            msum1 += a_m_b1 * a_m_b1;
-            d -= 8;
-        }
 
-        __m128 msum2 = _mm256_extractf128_ps(msum1, 1);
-        msum2 += _mm256_extractf128_ps(msum1, 0);
-
-        if (d >= 4) {
-            __m128 mx = _mm_loadu_ps(x);
-            x += 4;
-            __m128 my = _mm_loadu_ps(y);
-            y += 4;
-            const __m128 a_m_b1 = mx - my;
-            msum2 += a_m_b1 * a_m_b1;
-            d -= 4;
-        }
-
-        if (d > 0) {
-            __m128 mx = masked_read(d, x);
-            __m128 my = masked_read(d, y);
-            __m128 a_m_b1 = mx - my;
-            msum2 += a_m_b1 * a_m_b1;
-        }
-
-        msum2 = _mm_hadd_ps(msum2, msum2);
-        msum2 = _mm_hadd_ps(msum2, msum2);
-        return _mm_cvtss_f32(msum2);
         // return result;
+        return 0;
     }
 };
 
@@ -106,122 +61,23 @@ class DistanceInnerProduct : public Distance {
         // cannot use AVX2 _mm_mask_set1_epi32
     }
     float compare(const float *a, const float *b, unsigned size) const {
-
-#ifdef __GNUC__
-#ifdef __AVX__
-#define AVX_DOT(addr1, addr2, dest, tmp1, tmp2) \
-    tmp1 = _mm256_load_ps(addr1);               \
-    tmp2 = _mm256_load_ps(addr2);               \
-    tmp1 = _mm256_mul_ps(tmp1, tmp2);           \
-    dest = _mm256_add_ps(dest, tmp1);
-
-#else
-#ifdef __SSE2__
-#define SSE_DOT(addr1, addr2, dest, tmp1, tmp2) \
-    tmp1 = _mm128_loadu_ps(addr1);              \
-    tmp2 = _mm128_loadu_ps(addr2);              \
-    tmp1 = _mm128_mul_ps(tmp1, tmp2);           \
-    dest = _mm128_add_ps(dest, tmp1);
-        __m128 sum;
-        __m128 l0, l1, l2, l3;
-        __m128 r0, r1, r2, r3;
-        unsigned D = (size + 3) & ~3U;
-        unsigned DR = D % 16;
-        unsigned DD = D - DR;
-        const float *l = a;
-        const float *r = b;
-        const float *e_l = l + DD;
-        const float *e_r = r + DD;
-        float unpack[4] __attribute__((aligned(16))) = {0, 0, 0, 0};
-
-        sum = _mm_load_ps(unpack);
-        switch (DR) {
-            case 12:
-                SSE_DOT(e_l + 8, e_r + 8, sum, l2, r2);
-            case 8:
-                SSE_DOT(e_l + 4, e_r + 4, sum, l1, r1);
-            case 4:
-                SSE_DOT(e_l, e_r, sum, l0, r0);
-            default:
-                break;
-        }
-        for (unsigned i = 0; i < DD; i += 16, l += 16, r += 16) {
-            SSE_DOT(l, r, sum, l0, r0);
-            SSE_DOT(l + 4, r + 4, sum, l1, r1);
-            SSE_DOT(l + 8, r + 8, sum, l2, r2);
-            SSE_DOT(l + 12, r + 12, sum, l3, r3);
-        }
-        _mm_storeu_ps(unpack, sum);
-        result += unpack[0] + unpack[1] + unpack[2] + unpack[3];
-#else
-
-        float dot0, dot1, dot2, dot3;
-        const float *last = a + size;
-        const float *unroll_group = last - 3;
-
-        /* Process 4 items with each loop for efficiency. */
-        while (a < unroll_group) {
-            dot0 = a[0] * b[0];
-            dot1 = a[1] * b[1];
-            dot2 = a[2] * b[2];
-            dot3 = a[3] * b[3];
-            result += dot0 + dot1 + dot2 + dot3;
-            a += 4;
-            b += 4;
+        __m256 msum0 = _mm256_setzero_ps();
+        for (int i = 0; i < size; i += 8) {
+            __m256 mx = _mm256_loadu_ps(a + i);
+            __m256 my = _mm256_loadu_ps(b + i);
+            const __m256 a_m_b1 = _mm256_mul_ps(mx, my);
+            msum0 = _mm256_add_ps(msum0, a_m_b1);
         }
-        /* Process last 0-3 pixels.  Not needed for standard vector lengths. */
-        while (a < last) {
-            result += *a++ * *b++;
-        }
-#endif
-#endif
-#endif
-        // using avx-512
-        __m512 msum0 = _mm512_setzero_ps();
-
-        while (size >= 16) {
-            __m512 mx = _mm512_loadu_ps(a);
-            a += 16;
-            __m512 my = _mm512_loadu_ps(b);
-            b += 16;
-            msum0 = _mm512_add_ps(msum0, _mm512_mul_ps(mx, my));
-            size -= 16;
+        float result[8];
+        _mm256_storeu_ps(result, msum0);
+        float sum = 0;
+        for (int i = 0; i < 8; i++) {
+            sum += result[i];
         }
 
-        __m256 msum1 = _mm512_extractf32x8_ps(msum0, 1);
-        msum1 += _mm512_extractf32x8_ps(msum0, 0);
+        return -1.0 * sum;
 
-        if (size >= 8) {
-            __m256 mx = _mm256_loadu_ps(a);
-            a += 8;
-            __m256 my = _mm256_loadu_ps(b);
-            b += 8;
-            msum1 = _mm256_add_ps(msum1, _mm256_mul_ps(mx, my));
-            size -= 8;
-        }
 
-        __m128 msum2 = _mm256_extractf128_ps(msum1, 1);
-        msum2 += _mm256_extractf128_ps(msum1, 0);
-
-        if (size >= 4) {
-            __m128 mx = _mm_loadu_ps(a);
-            a += 4;
-            __m128 my = _mm_loadu_ps(b);
-            b += 4;
-            msum2 = _mm_add_ps(msum2, _mm_mul_ps(mx, my));
-            size -= 4;
-        }
-
-        if (size > 0) {
-            __m128 mx = masked_read(size, a);
-            __m128 my = masked_read(size, b);
-            msum2 = _mm_add_ps(msum2, _mm_mul_ps(mx, my));
-        }
-
-        msum2 = _mm_hadd_ps(msum2, msum2);
-        msum2 = _mm_hadd_ps(msum2, msum2);
-        return -1.0 * _mm_cvtss_f32(msum2);
-        // return result;
     }
 };
 class DistanceFastL2 : public DistanceInnerProduct {
diff --git a/include/index_bipartite.h b/include/index_bipartite.h
index 802f3fd..541f3f2 100644
--- a/include/index_bipartite.h
+++ b/include/index_bipartite.h
@@ -37,6 +37,8 @@ class IndexBipartite : public Index {
 
     void BuildRoarGraph(size_t n_sq, const float *sq_data, size_t n_bp, const float *bp_data,
                                 const Parameters &parameters);
+    void BuildRoarGraphwithData(size_t n_sq, const float *sq_data, size_t n_bp, const float *bp_data,
+                                const Parameters &parameters);
     virtual void Build(size_t n, const float *data, const Parameters &parameters) override;
 
     inline void SetBipartiteParameters(const Parameters &parameters) {}
@@ -108,6 +110,11 @@ class IndexBipartite : public Index {
 
     void LoadLearnBaseKNN(const char *filename);
 
+    void SetLearnBaseKNN(const uint32_t* learn_base_knn, uint32_t npts, uint32_t k_dim);
+
+    uint32_t SearchRoarGraphPy(const float *query, size_t k, size_t &qid, uint32_t L_pq,
+                                               unsigned *indices, float* res_dists);
+
     void LoadBaseLearnKNN(const char *filename);
 
     inline std::vector<std::vector<uint32_t>> &GetProjectionGraph() { return projection_graph_; }
diff --git a/pyroar/index_bindings.cpp b/pyroar/index_bindings.cpp
new file mode 100644
index 0000000..562f455
--- /dev/null
+++ b/pyroar/index_bindings.cpp
@@ -0,0 +1,112 @@
+
+#include <omp.h>
+#include <pybind11/functional.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <algorithm>
+#include <boost/dynamic_bitset.hpp>
+#include <boost/program_options.hpp>
+#include <chrono>
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <filesystem>
+#include <unistd.h>
+
+#include <faiss/IndexFlat.h>
+#include <faiss/IndexIVFPQ.h>
+#include <faiss/IndexIVFFlat.h>
+#include <faiss/AutoTune.h>
+#include <faiss/index_factory.h>
+
+#include "index_bipartite.h"
+#include "efanna2e/distance.h"
+#include "efanna2e/neighbor.h"
+#include "efanna2e/parameters.h"
+#include "efanna2e/util.h"
+
+namespace py = pybind11;
+
+template <typename T>
+
+class IndexRoarGraph {
+   public:
+    IndexRoarGraph(const size_t dimension, const size_t n, efanna2e::Metric m) {
+        index_ = new efanna2e::IndexBipartite(dimension, n, m, nullptr);
+    }
+
+    ~IndexRoarGraph() { delete index_; }
+
+
+    void setThreads(uint32_t num_threads) {
+        omp_set_num_threads(num_threads);
+        index_->InitVisitedListPool(2 * num_threads);
+    }
+
+    void buildIndex(uint32_t sq_num, uint32_t k_dim, uint32_t base_num, uint32_t M_sq, uint32_t M_pjbp, 
+                    uint32_t L_pjpq, uint32_t num_threads, py::array_t<uint32_t, py::array::c_style | py::array::forcecast> &learn_base_nn, py::array_t<float, py::array::c_style | py::array::forcecast> &base_data) {
+        efanna2e::Parameters parameters;
+        parameters.Set<uint32_t>("M_sq", M_sq);
+        parameters.Set<uint32_t>("M_pjbp", M_pjbp);
+        parameters.Set<uint32_t>("L_pjpq", L_pjpq);
+        parameters.Set<uint32_t>("num_threads", num_threads);
+        const uint32_t* learn_base_nn_ptr = &(learn_base_nn.unchecked()(0, 0));
+        const float* base_data_ptr = &(base_data.unchecked()(0, 0));
+        index_->SetLearnBaseKNN(learn_base_nn_ptr, sq_num, k_dim);
+        index_->BuildRoarGraphwithData(sq_num, nullptr, base_num, base_data_ptr, parameters);
+    }
+
+    void searchIndex(py::array_t<float, py::array::c_style | py::array::forcecast> &q_data, size_t k, uint32_t L_pq,
+                py::array_t<uint32_t, py::array::c_style> &res_id, py::array_t<float, py::array::c_style> &res_dist,
+                size_t q_num, uint32_t num_threads, bool using_sq = false) {
+        auto items = q_data.unchecked();
+        auto res = res_id.mutable_unchecked();
+        auto dists = res_dist.mutable_unchecked();
+#pragma omp parallel for schedule(dynamic, 1)
+            for (size_t qid = 0; qid < q_num; qid++) {
+                const float *query = &items(qid, 0);
+                uint32_t cmps = index_->SearchRoarGraphPy(query, k, qid, L_pq, &res(qid, 0), &dists(qid, 0));
+                #pragma omp atomic
+                total_cmps += cmps;
+            }
+    }
+
+    void reset_cmps() {
+        total_cmps = 0;
+    }
+
+    uint32_t get_cmps() {
+        return total_cmps;
+    }
+
+    void saveProjectionGraph(const std::string& filename) {
+        index_->SaveProjectionGraph(filename.c_str());
+    }
+
+   private:
+    efanna2e::IndexBipartite *index_;
+    uint32_t total_cmps = 0;
+};
+
+
+PYBIND11_MODULE(RoarGraph, m) {
+    m.doc() = "pybind11 RoarGraph plugin";  // optional module docstring
+    py::enum_<efanna2e::Metric>(m, "Metric")
+        .value("L2", efanna2e::Metric::L2)
+        .value("IP", efanna2e::Metric::INNER_PRODUCT)
+        .value("COSINE", efanna2e::Metric::COSINE)
+        .export_values();
+    py::class_<IndexRoarGraph<float>>(m, "IndexRoarGraph")
+        .def(py::init<const size_t, const size_t, efanna2e::Metric>())
+        .def("search", &IndexRoarGraph<float>::searchIndex)
+        .def("build", &IndexRoarGraph<float>::buildIndex)
+        .def("setThreads", &IndexRoarGraph<float>::setThreads)
+        .def("reset_cmps", &IndexRoarGraph<float>::reset_cmps)
+        .def("get_cmps", &IndexRoarGraph<float>::get_cmps)
+        .def("save_projection_graph", &IndexRoarGraph<float>::saveProjectionGraph, py::arg("filename"));
+}
\ No newline at end of file
diff --git a/pyroar/setup.py b/pyroar/setup.py
new file mode 100644
index 0000000..91b8545
--- /dev/null
+++ b/pyroar/setup.py
@@ -0,0 +1,87 @@
+from glob import glob
+from setuptools import setup, Extension
+import numpy as np
+import pybind11
+from setuptools.command.build_ext import build_ext
+import setuptools
+import os
+import sys
+from pybind11.setup_helpers import Pybind11Extension, build_ext
+
+include_dirs = [
+    # Path to pybind11 headers
+    pybind11.get_include(),
+    np.get_include()
+]
+
+include_dirs.extend(['../include/', '../include/efanna2e', '../thirdparty/robin-map/include'])
+
+module = Pybind11Extension('RoarGraph',
+             sources=['index_bindings.cpp',
+                      '../src/index.cpp', '../src/index_bipartite.cpp'
+                      ],
+             include_dirs=include_dirs,
+            # Prefer generic FAISS runtime library for wider CPU compatibility.
+             libraries=['faiss'],
+             language='c++')
+
+def has_flag(compiler, flagname):
+    """Return a boolean indicating whether a flag name is supported on
+    the specified compiler.
+    """
+    import tempfile
+    with tempfile.NamedTemporaryFile('w', suffix='.cpp') as f:
+        f.write('int main (int argc, char **argv) { return 0; }')
+        try:
+            compiler.compile([f.name], extra_postargs=[flagname])
+        except setuptools.distutils.errors.CompileError:
+            return False
+    return True
+
+def cpp_flag(compiler):
+    """Return the -std=c++[11/14] compiler flag.
+    The c++14 is prefered over c++11 (when it is available).
+    """
+
+    if has_flag(compiler, '-std=c++20'):
+        return '-std=c++20'
+    elif has_flag(compiler, '-std=c++17'):
+        return '-std=c++17'
+    else:
+        raise RuntimeError('Unsupported compiler -- at least C++11 support '
+                           'is needed!')
+
+class BuildExt(build_ext):
+    """A custom build extension for adding compiler-specific options."""
+    c_opts = {
+        # Keep baseline flags portable for heterogeneous Linux hosts.
+        # Aggressive CPU-specific flags like -march=native/-mavx2 can produce
+        # binaries that crash with "Illegal instruction" on older CPUs.
+        'unix': ['-O3', '-fopenmp', '-DNDEBUG', '-ftree-vectorize', '-Wall', '-DINFO', '-mavx'],
+    }
+
+    link_opts = {
+        'unix': [],
+    }
+
+    c_opts['unix'].append("-fopenmp")
+    link_opts['unix'].extend(['-fopenmp', '-pthread'])
+
+    def build_extensions(self):
+        ct = self.compiler.compiler_type
+        opts = self.c_opts.get(ct, [])
+        opts.append(cpp_flag(self.compiler))
+
+        for ext in self.extensions:
+            ext.extra_compile_args.extend(opts)
+            ext.extra_link_args.extend(self.link_opts.get(ct, []))
+
+        build_ext.build_extensions(self)
+
+setup(name='RoarGraph',
+    version='1.0',
+    description='Python bindings for RoarGraph class',
+    ext_modules=[module],
+    cmdclass={'build_ext': BuildExt},
+    install_requires=['numpy', 'pybind11']
+    )
\ No newline at end of file
diff --git a/src/index.cpp b/src/index.cpp
index 78a7410..a042d09 100644
--- a/src/index.cpp
+++ b/src/index.cpp
@@ -14,9 +14,10 @@ Index::Index(const size_t dimension, const size_t n, Metric metric = L2)
             std::cout << "Inside using L2 distance." << std::endl;
             break;
         case efanna2e::COSINE:
+            distance_ = new DistanceInnerProduct();
+            break;
         case efanna2e::INNER_PRODUCT:
             distance_ = new DistanceInnerProduct();
-            std::cout << "Inside using IP distance after normalization." << std::endl;
             break;
         default:
             distance_ = new DistanceL2();
diff --git a/src/index_bipartite.cpp b/src/index_bipartite.cpp
index cc7d2d6..d24ef79 100644
--- a/src/index_bipartite.cpp
+++ b/src/index_bipartite.cpp
@@ -30,9 +30,9 @@ namespace efanna2e {
 IndexBipartite::IndexBipartite(const size_t dimension, const size_t n, Metric m, Index *initializer)
     : Index(dimension, n, m), initializer_{initializer}, total_pts_const_(n) {
     bipartite_ = true;
-    l2_distance_ = new DistanceL2();
     width_ = 1;
     if (m == efanna2e::COSINE) {
+        std::cout << "Inside using IP distance after normalization." << std::endl;
         need_normalize = true;
     }
 }
@@ -205,15 +205,7 @@ void IndexBipartite::BuildRoarGraph(size_t n_sq, const float *sq_data, size_t n_
     assert(projection_ep_ < nd_);
     std::cout << "begin link projection" << std::endl;
     LinkProjection(parameters);
-    std::cout << std::endl;
-    // std::cout << "Starting collect points" << std::endl;
-    // auto co_s = std::chrono::high_resolution_clock::now();
-    // CollectPoints(parameters);
-    // auto co_e = std::chrono::high_resolution_clock::now();
-    // diff = co_e - co_s;
-    // std::cout << "Collect points time: " << diff.count() << std::endl;
 
-    // e = std::chrono::high_resolution_clock::now();
     auto e = std::chrono::high_resolution_clock::now();
     auto diff = e - s;
     std::cout << "Build projection graph time: " << diff.count() / (1000 * 1000 * 1000) << std::endl;
@@ -231,6 +223,65 @@ void IndexBipartite::BuildRoarGraph(size_t n_sq, const float *sq_data, size_t n_
 
     has_built = true;
 }
+void IndexBipartite::BuildRoarGraphwithData(size_t n_sq, const float *sq_data, size_t n_bp, const float *bp_data,
+                                    const Parameters &parameters) {
+    auto s = std::chrono::high_resolution_clock::now();
+    uint32_t M_sq = parameters.Get<uint32_t>("M_sq");
+    uint32_t M_pjbp = parameters.Get<uint32_t>("M_pjbp");
+
+    float *bp_data_copy = new float[n_bp * dimension_];
+    memcpy(bp_data_copy, bp_data, n_bp * dimension_ * sizeof(float));
+    data_bp_ = bp_data_copy;
+    data_sq_ = sq_data;
+    nd_ = n_bp;
+    nd_sq_ = n_sq;
+    total_pts_ = nd_ + nd_sq_;
+    u32_nd_ = static_cast<uint32_t>(nd_);
+    u32_nd_sq_ = static_cast<uint32_t>(nd_sq_);
+    u32_total_pts_ = static_cast<uint32_t>(total_pts_);
+    locks_ = std::vector<std::mutex>(total_pts_);
+
+    SetBipartiteParameters(parameters);
+
+    if (need_normalize) {
+        std::cout << "normalizing base data" << std::endl;
+        for (size_t i = 0; i < nd_; ++i) {
+            float *data = const_cast<float *>(data_bp_);
+            normalize(data + i * dimension_, dimension_);
+        }
+    }
+
+
+    float bipartite_degree_avg = 0;
+    size_t bipartite_degree_max = 0, bipartite_degree_min = std::numeric_limits<size_t>::max();
+    float projection_degree_avg = 0;
+    size_t projection_degree_max = 0, projection_degree_min = std::numeric_limits<size_t>::max();
+
+
+    size_t i = 0;
+
+    supply_nbrs_.resize(nd_);
+
+    BipartiteProjectionReserveSpace(parameters);
+
+    CalculateProjectionep();
+
+    assert(projection_ep_ < nd_);
+    LinkProjection(parameters);
+
+    auto e = std::chrono::high_resolution_clock::now();
+    auto diff = e - s;
+
+    for (i = 0; i < projection_graph_.size(); ++i) {
+        std::vector<uint32_t> &nbrs = projection_graph_[i];
+        projection_degree_avg += static_cast<float>(nbrs.size());
+        projection_degree_max = std::max(projection_degree_max, nbrs.size());
+        projection_degree_min = std::min(projection_degree_min, nbrs.size());
+    }
+
+    has_built = true;
+}
+
 
 void IndexBipartite::qbaseNNbipartite(const Parameters &parameters) {
     uint32_t M_pjbp = parameters.Get<uint32_t>("M_pjbp");
@@ -949,7 +1000,6 @@ void IndexBipartite::PruneLocalJoinCandidates(uint32_t node, const Parameters &p
 }
 
 void IndexBipartite::BipartiteProjectionReserveSpace(const Parameters &parameters) {
-    std::cout << "begin projection graph init" << std::endl;
     uint32_t M_pjbp = parameters.Get<uint32_t>("M_pjbp");
     projection_graph_.resize(u32_nd_);
     for (uint32_t i = 0; i < u32_nd_; ++i) {
@@ -970,7 +1020,7 @@ void IndexBipartite::TrainingLink2Projection(const Parameters &parameters, Simpl
         vis_order_sq.push_back(i);
     }
 
-    #pragma omp parallel for schedule(static, 100)
+    #pragma omp parallel for schedule(dynamic, 100)
     for (uint32_t it_sq = 0; it_sq < u32_nd_sq_; ++it_sq) {
         uint32_t sq = vis_order_sq[it_sq];
         // boost::dynamic_bitset<> visited{u32_total_pts_, false};
@@ -1008,7 +1058,7 @@ void IndexBipartite::TrainingLink2Projection(const Parameters &parameters, Simpl
     std::cout << std::endl;
     std::atomic<uint32_t> degree_cnt(0);
     std::atomic<uint32_t> zero_cnt(0);
-#pragma omp parallel for schedule(static, 100)
+#pragma omp parallel for schedule(dynamic, 100)
     for (uint32_t i = 0; i < vis_order.size(); ++i) {
         uint32_t node = vis_order[i];
         if (projection_graph_[node].size() < M_pjbp) {
@@ -1056,7 +1106,7 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
     }
     std::chrono::high_resolution_clock::time_point t1 = std::chrono::high_resolution_clock::now();
 
-#pragma omp parallel for schedule(static, 100)
+#pragma omp parallel for schedule(dynamic, 100)
     for (uint32_t it_sq = 0; it_sq < u32_nd_sq_; ++it_sq) {
         uint32_t sq = vis_order_sq[it_sq];
 
@@ -1066,18 +1116,13 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
             nn_base.shrink_to_fit();
         }
         uint32_t choose_tgt = 0;
-        // for (size_t i = 0; i < 100; ++i) {
-        //     if (projection_graph_[nn_base[i]].size() < M_pjbp) {
-        //         choose_tgt = nn_base[i];
-        //         break;
-        //     }
-        // }
         uint32_t cur_tgt = nn_base[choose_tgt];
         std::vector<Neighbor> full_retset;
         for (size_t i = 0; i < nn_base.size(); ++i) {
             if (nn_base[i] == cur_tgt) {
                 continue;
             }
+            _mm_prefetch((const char *)(data_bp_ + dimension_ * (uint64_t)nn_base[i]), _MM_HINT_T0);
             float distance = distance_->compare(data_bp_ + dimension_ * (uint64_t)nn_base[i], data_bp_ + dimension_ * (uint64_t)cur_tgt,
                                                 (unsigned)dimension_);
             full_retset.push_back(Neighbor(nn_base[i], distance, false));
@@ -1090,21 +1135,10 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
             projection_graph_[cur_tgt] = pruned_list;
         }
         ProjectionAddReverse(cur_tgt, parameters);
-        if (sq % 1000 == 0) {
-            std::cout << "\r" << (100.0 * sq) / u32_nd_sq_ << "% of projection search bipartite by base completed."
-                      << std::flush;
-        }
-    }
-
-    std::cout << std::endl;
-#pragma omp parallel for schedule(static, 100)
-    for (uint32_t i = 0; i < vis_order.size(); ++i) {
-        uint32_t node = vis_order[i];
-        ProjectionAddReverse(node, parameters);
     }
 
 
-#pragma omp parallel for schedule(static, 2048)
+#pragma omp parallel for schedule(dynamic, 100)
     for (uint32_t i = 0; i < vis_order.size(); ++i) {
         size_t node = (size_t)vis_order[i];
         if (projection_graph_[node].size() > M_pjbp) {
@@ -1129,7 +1163,6 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
             PruneBiSearchBaseGetBase(full_retset, data_bp_ + dimension_ * (size_t)node, node, parameters, prune_list);
             {
                 LockGuard guard(locks_[node]);
-                projection_graph_[node].clear();
                 projection_graph_[node] = prune_list;
             }
         }
@@ -1137,59 +1170,17 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
 
     std::chrono::high_resolution_clock::time_point t2 = std::chrono::high_resolution_clock::now();
 
-    // save t2 - t1 in seconds in projection time
     auto projection_time = std::chrono::duration_cast<std::chrono::duration<double>>(t2 - t1).count();
 
-    std::atomic<uint32_t> degree_cnt(0);
-    std::atomic<uint32_t> zero_cnt(0);
-#pragma omp parallel for schedule(static, 100)
-    for (uint32_t i = 0; i < vis_order.size(); ++i) {
-        uint32_t node = vis_order[i];
-        if (projection_graph_[node].size() < M_pjbp) {
-            // std::cout << "Warning: projection graph node " << node << " has less than M_pjbp neighbors." << std::endl;
-            degree_cnt.fetch_add(1);
-            if (projection_graph_[node].size() == 0) {
-                zero_cnt.fetch_add(1);
-            }
-        }
-    }
-    std::cout << "Projection time: " << projection_time << std::endl;
-    std::cout << "Warning: " << degree_cnt.load() << " nodes have less than M_pjbp neighbors." << std::endl;
-    std::cout << "Warning: " << zero_cnt.load() << " nodes have no neighbors." << std::endl;
-
-    // stats projection graph degree
-    float avg_degree = 0;
-    uint64_t total_degree = 0;
-    uint32_t max_degree = 0;
-    uint32_t min_degree = std::numeric_limits<uint32_t>::max();
-    for (uint32_t i = 0; i < u32_nd_; ++i) {
-        if (projection_graph_[i].size() > max_degree) {
-            max_degree = projection_graph_[i].size();
-        }
-        if (projection_graph_[i].size() < min_degree) {
-            min_degree = projection_graph_[i].size();
-        }
-        avg_degree += static_cast<float>(projection_graph_[i].size());
-        total_degree += projection_graph_[i].size();
-    }
-    std::cout << "total degree: " << total_degree << std::endl;
-    avg_degree /= (float)u32_nd_;
-    std::cout << "After projection, average degree of projection graph: " << avg_degree << std::endl;
-    std::cout << "After projection, max degree of projection graph: " << max_degree << std::endl;
-    std::cout << "After projection, min degree of projection graph: " << min_degree << std::endl;
-
-    std::cout << std::endl;
 
     for (size_t i = 0; i < projection_graph_.size(); ++i) {
-        // projection_graph_[i].clear();
         supply_nbrs_[i] = projection_graph_[i];
         supply_nbrs_[i].reserve(M_pjbp * 2 * PROJECTION_SLACK);
-        // supply_nbrs_[i].reserve(M_pjbp * PROJECTION_SLACK);
     }
 
     t1 = std::chrono::high_resolution_clock::now();
 
-#pragma omp parallel for schedule(dynamic, 2048)
+#pragma omp parallel for schedule(dynamic, 100)
     for (uint32_t i = 0; i < nd_; ++i) {
         size_t node = vis_order[i];
         boost::dynamic_bitset<> visited{u32_nd_, false};
@@ -1213,40 +1204,8 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
             supply_nbrs_[node] = pruned_list;
         }
         SupplyAddReverse(node, parameters);
-        if (node % 1000 == 0) {
-            std::cout << "\r" << (100.0 * node) / (u32_nd_) << "% of projection graph base search completed."
-                      << std::flush;
-        }
     }
 
-    std::cout << "finish connectivity enhancement" << std::endl;
-
-#pragma omp parallel for schedule(dynamic, 2048)
-    for (uint32_t i = 0; i < nd_; ++i) {
-        size_t node = vis_order[i];
-        if (supply_nbrs_[node].size() > M_pjbp) {
-            std::vector<Neighbor> full_retset;
-            tsl::robin_set<uint32_t> visited;
-            for (size_t j = 0; j < supply_nbrs_[node].size(); ++j) {
-                if (visited.find(supply_nbrs_[node][j]) != visited.end()) {
-                    continue;
-                }
-                float distance = distance_->compare(data_bp_ + dimension_ * supply_nbrs_[node][j],
-                                                    data_bp_ + dimension_ * node, dimension_);
-                visited.insert(supply_nbrs_[node][j]);
-                full_retset.push_back(Neighbor(supply_nbrs_[node][j], distance, false));
-            }
-            std::vector<uint32_t> prune_list;
-            PruneProjectionBaseSearchCandidates(full_retset, data_bp_ + dimension_ * node, node, parameters,
-                                                prune_list);
-            {
-                LockGuard guard(locks_[node]);
-                supply_nbrs_[node].clear();
-                supply_nbrs_[node] = prune_list;
-            }
-        }
-    }
-    std::cout << "finish connectivity enhancement degree check" << std::endl;
 
 #pragma omp parallel for schedule(dynamic, 100)
     for (size_t i = 0; i < projection_graph_.size(); ++i) {
@@ -1262,20 +1221,14 @@ void IndexBipartite::LinkProjection(const Parameters &parameters) {
             }
         }
         projection_graph_[i].insert(projection_graph_[i].end(), ok_insert.begin(), ok_insert.end());
-        // projection_graph_[i] = ok_insert;
-        // std::copy(ok_insert.begin(), ok_insert.end(), projection_graph_[i].begin() + M_pjbp);
-        // std::copy(ok_insert.begin(), ok_insert.end(), projection_graph_[i].begin() + projection_graph_[i].size());
-        // std::copy(ok_insert.begin(), ok_insert.end(), projection_graph_[i].begin());
     }
 
     t2 = std::chrono::high_resolution_clock::now();
 
-    // save t2 - t1 in seconds in connectivity enhancement time
-    auto connectivity_enhancement_time = std::chrono::duration_cast<std::chrono::duration<double>>(t2 - t1).count();
 
-    std::cout << "Connectivity enhancement time: " << connectivity_enhancement_time << std::endl;
 }
 
+
 void IndexBipartite::SearchProjectionGraphInternal(NeighborPriorityQueue &search_queue, const float *query,
                                                    uint32_t tgt, const Parameters &parameters,
                                                    boost::dynamic_bitset<> &visited,
@@ -1484,36 +1437,6 @@ void IndexBipartite::PruneProjectionInternalReverseCandidates(uint32_t src_node,
             }
         }
     }
-    start = 0;
-    while (result.size() < M_pjbp && (++start) < prune_queue.size()) {
-        auto &p = prune_queue[start];
-        bool occlude = false;
-        for (size_t i = 0; i < result.size(); ++i) {
-            if (p.id == result[i]) {
-                occlude = true;
-                break;
-            }
-            // efanna2e::prefetch_vector((char *)(data_bp_ + p.id * dimension_), dimension_ * sizeof(float));
-            float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[i],
-                                           (unsigned)dimension_);
-            // float djk = l2_distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[i],
-            //                                   (unsigned)dimension_);
-            // if (metric_ == efanna2e::Metric::INNER_PRODUCT) {
-            //     djk = -djk;
-            // }
-            if (1.0 * djk < p.distance) {
-                occlude = true;
-                break;
-            }
-        }
-        if (!occlude) {
-            if (p.id != src_node) {
-                if (std::find(result.begin(), result.end(), p.id) == result.end()) {
-                    result.push_back(p.id);
-                }
-            }
-        }
-    }
 
     // for (size_t i = 0; i < pruned_list.size() && result.size() < M_pjbp; ++i) {
     //     if (std::find(result.begin(), result.end(), pruned_list[i]) == result.end()) {
@@ -1556,7 +1479,7 @@ void IndexBipartite::PruneProjectionReverseCandidates(uint32_t src_node, const P
             }
             float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[i],
                                            (unsigned)dimension_);
-            if (djk < p.distance) {
+            if (1.3 * djk < p.distance) {
                 occlude = true;
                 break;
             }
@@ -1568,30 +1491,6 @@ void IndexBipartite::PruneProjectionReverseCandidates(uint32_t src_node, const P
         }
     }
 
-    start = 0;
-    while (result.size() < M_pjbp && (++start) < prune_queue.size()) {
-        auto &p = prune_queue[start];
-        bool occlude = false;
-        for (size_t i = 0; i < result.size(); ++i) {
-            if (p.id == result[i]) {
-                occlude = true;
-                break;
-            }
-            float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[i],
-                                           (unsigned)dimension_);
-            if (1.0 * djk < p.distance) {
-                occlude = true;
-                break;
-            }
-        }
-        if (!occlude) {
-            if (p.id != src_node) {
-                if (std::find(result.begin(), result.end(), p.id) == result.end()) {
-                    result.push_back(p.id);
-                }
-            }
-        }
-    }
 
     for (size_t i = 0; i < pruned_list.size() && result.size() < M_pjbp; ++i) {
         if (std::find(result.begin(), result.end(), pruned_list[i]) == result.end()) {
@@ -1643,7 +1542,7 @@ void IndexBipartite::PruneBiSearchBaseGetBase(std::vector<Neighbor> &search_pool
                 break;
             }
             float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[t], dimension_);
-            if (djk < p.distance) {
+            if (1.3 * djk < p.distance) {
                 occlude = true;
                 break;
             }
@@ -1655,32 +1554,6 @@ void IndexBipartite::PruneBiSearchBaseGetBase(std::vector<Neighbor> &search_pool
         }
     }
 
-    start = 0;
-    while (result.size() < M_pjbp && (++start) < search_pool.size()) {
-        Neighbor &p = search_pool[start];
-        if (std::find(result.begin(), result.end(), p.id) != result.end()) {
-            continue;
-        }
-        bool occlude = false;
-        for (size_t t = 0; t < result.size(); ++t) {
-            if (p.id == result[t]) {
-                occlude = true;
-                break;
-            }
-            float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[t], dimension_);
-            if (1.0 * djk < p.distance) {
-                occlude = true;
-                break;
-            }
-        }
-        if (!occlude) {
-            if (p.id != tgt_base) {
-                if (std::find(result.begin(), result.end(), p.id) == result.end()) {
-                    result.push_back(p.id);
-                }
-            }
-        }
-    }
 
     for (size_t i = 1; i < base_pool.size() && result.size() < M_pjbp; ++i) {
         if (std::find(result.begin(), result.end(), base_pool[i].id) == result.end()) {
@@ -1880,7 +1753,7 @@ void IndexBipartite::PruneProjectionBaseSearchCandidates(std::vector<Neighbor> &
             // dimension_); if (metric_ == efanna2e::Metric::INNER_PRODUCT) {
             //     dik = -dik;
             // }
-            if (djk < p.distance) {
+            if (1.3 * djk < p.distance) {
                 occlude = true;
                 break;
             }
@@ -1893,39 +1766,6 @@ void IndexBipartite::PruneProjectionBaseSearchCandidates(std::vector<Neighbor> &
             }
         }
     }
-    start = 0;
-    while (result.size() < M_pjbp && (++start) < search_pool.size()) {
-        Neighbor &p = search_pool[start];
-        bool occlude = false;
-        for (size_t t = 0; t < result.size(); ++t) {
-            if (p.id == result[t]) {
-                occlude = true;
-                break;
-            }
-            float djk = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * result[t], dimension_);
-            // if (metric_ == efanna2e::Metric::INNER_PRODUCT) {
-            //     djk = -djk;
-            // }
-            // float dik = distance_->compare(data_bp_ + dimension_ * p.id, data_bp_ + dimension_ * src_node,
-            // dimension_); if (metric_ == efanna2e::Metric::INNER_PRODUCT) {
-            //     dik = -dik;
-            // }
-            if (1.0 * djk < p.distance) {
-                occlude = true;
-                break;
-            }
-        }
-        if (!occlude) {
-            if (p.id != qid) {
-                // if (std::find(src_nbrs.begin(), src_nbrs.end(), p.id) == src_nbrs.end()) {
-                if (std::find(result.begin(), result.end(), p.id) == result.end()) {
-                    result.push_back(p.id);
-                }
-                // result.push_back(p.id);
-                // }
-            }
-        }
-    }
 
     // for (size_t i = 0; i < search_pool.size() && result.size() < M_pjbp; ++i) {
     //     if (std::find(result.begin(), result.end(), search_pool[i].id) == result.end()) {
@@ -2035,7 +1875,6 @@ void IndexBipartite::CalculateProjectionep() {
     }
     projection_ep_ = closest;
     // projection_ep_ = min_idx;
-    std::cout << "projection ep: " << projection_ep_ << std::endl;
     delete[] center;
     delete[] distances;
 }
@@ -2419,6 +2258,60 @@ std::pair<uint32_t, uint32_t> IndexBipartite::SearchRoarGraph(const float *query
     return std::make_pair(cmps, hops);
 }
 
+uint32_t IndexBipartite::SearchRoarGraphPy(const float *query, size_t k, size_t &qid, uint32_t L_pq,
+                                               unsigned *indices, float* res_dists) {
+    NeighborPriorityQueue search_queue(L_pq);
+    std::vector<uint32_t> init_ids;
+    init_ids.push_back(projection_ep_);
+    prefetch_vector((char *)(data_bp_ + projection_ep_ * dimension_), dimension_);
+    VisitedList *vl = visited_list_pool_->getFreeVisitedList();
+    vl_type *visited_array = vl->mass;
+    vl_type visited_array_tag = vl->curV;
+
+    for (auto &id : init_ids) {
+
+        float distance = distance_->compare(data_bp_ + id * dimension_, query, (unsigned)dimension_);
+
+        Neighbor nn = Neighbor(id, distance, false);
+        search_queue.insert(nn);
+    }
+    uint32_t cmps = 0;
+    uint32_t hops = 0;
+    while (search_queue.has_unexpanded_node()) {
+        auto cur_check_node = search_queue.closest_unexpanded();
+        auto cur_id = cur_check_node.id;
+        uint32_t *cur_nbrs = projection_graph_[cur_id].data();
+
+        ++hops;
+        for (size_t j = 0; j < projection_graph_[cur_id].size(); ++j) {  // current check node's neighbors
+            uint32_t nbr = *(cur_nbrs + j);
+            _mm_prefetch((char *)(visited_array + *(cur_nbrs + j + 1)), _MM_HINT_T0);
+            _mm_prefetch((char *)(data_bp_ + *(cur_nbrs + j + 1) * dimension_), _MM_HINT_T0);
+            if (visited_array[nbr] != visited_array_tag) {
+                visited_array[nbr] = visited_array_tag;
+                float distance = distance_->compare(data_bp_ + nbr * dimension_, query, (unsigned)dimension_);
+
+
+                ++cmps;
+                search_queue.insert({nbr, distance, false});
+            }
+        }
+    }
+    visited_list_pool_->releaseVisitedList(vl);
+
+    if (unlikely(search_queue.size() < k)) {
+        std::stringstream ss;
+        ss << "not enough results: " << search_queue.size() << ", expected: " << k;
+        throw std::runtime_error(ss.str());
+    }
+
+    for (size_t i = 0; i < k; ++i) {
+        indices[i] = search_queue[i].id;
+        res_dists[i] = search_queue[i].distance;
+    }
+    return cmps;
+}
+
 // uint32_t IndexBipartite::SearchProjectionGraph(const float *query, size_t k, size_t &qid, const Parameters
 // &parameters,
 //                                                unsigned *indices) {
@@ -2638,6 +2531,16 @@ void IndexBipartite::LoadLearnBaseKNN(const char *filename) {
     in.close();
 }
 
+void IndexBipartite::SetLearnBaseKNN(const uint32_t* learn_base_knn, uint32_t npts, uint32_t k_dim) {
+    learn_base_knn_.resize(npts);
+#pragma omp parallel for
+    for (uint32_t i = 0; i < npts; i++) {
+        learn_base_knn_[i].resize(k_dim);
+        memcpy(learn_base_knn_[i].data(), learn_base_knn + i * k_dim, sizeof(uint32_t) * k_dim);
+
+    }
+}
+
 void IndexBipartite::LoadBaseLearnKNN(const char *filename) {
     std::ifstream in(filename, std::ios::binary);
     uint32_t npts;
